File: src/commands/user/devicedelete.ts
import {UserDeviceListResponse} from "@ironcorelabs/ironnode";
import {Command, Flags} from "@oclif/core";
import {ironnode} from "../../lib/SDK";
import {keyFile} from "../../lib/sharedFlags";
import {buildCommandSampleText, ErrorOr, isError} from "../../lib/Utils";
import chalk = require("chalk");

type DeviceDeleteResult = ErrorOr<{id: number}>;

/**
 * Delete one to many devices given their ID. Verifies that the user isn't trying to use this command to delete the device they're on and makes requests
 * in parallel to delete devices for each ID provided.
 */
export default class DeviceDelete extends Command {
    static aliases = ["user:deviceDelete"];
    static description =
        "Deauthorize a device from being able to decrypt your data. Use this to deauthorize devices other than the device you are currently on. To deauthorize the current device, use 'ironhide logout'.";
    static args = [
        {
            name: "device ID",
            description: "ID of the device keys to revoke. Multiple IDs can be provided.",
            required: true,
        },
    ];
    static strict = false;
    static flags = {
        help: Flags.help({char: "h"}),
        keyfile: keyFile(),
    };
    static examples = [buildCommandSampleText(`user:devicedelete 293 5559 312`)];

    /**
     * Return a Promise operation to delete the provided device ID. Maps any failures into successes so we can display information about each
     * result back to the user.
     */
    getDeviceDeletePromise(deviceID: string): Promise<DeviceDeleteResult> {
        const numericalDeviceID = parseInt(deviceID);
        if (!numericalDeviceID) {
            return Promise.resolve(new Error(`Expected a numerical device ID but got '${deviceID}' instead.`));
        }
        return ironnode()
            .user.deleteDevice(numericalDeviceID)
            .catch((e) => e);
    }

    /**
     * Check if the user is attempting to delete the device keys that they're currently using. If so display an error.
     */
    async checkDeletionOfCurrentDevice(deviceIDs: string[]) {
        let devices: UserDeviceListResponse;
        try {
            devices = await ironnode().user.listDevices();
        } catch (e) {
            throw new Error("Failed to make request to get current device keys.");
        }
        const currentDevice = devices.result.find((device) => device.isCurrentDevice);
        if (currentDevice && deviceIDs.includes(`${currentDevice.id}`)) {
            throw new Error("Attempting to delete keys for the current device. Use the 'logout' command instead.");
        }
    }

    async run() {
        const {argv} = await this.parse(DeviceDelete);
        try {
            await this.checkDeletionOfCurrentDevice(argv);
        } catch (e: any) {
            return this.error(chalk.red(e.message));
        }

        const deviceDeletes = argv.map((deviceID) => this.getDeviceDeletePromise(deviceID));
        return Promise.all(deviceDeletes).then((resultList) => {
            let successfulCount = 0;
            let failureCount = 0;
            resultList.forEach((deviceDeleteResult) => {
                if (isError(deviceDeleteResult)) {
                    this.log(chalk.red(deviceDeleteResult.message));
                    failureCount++;
                } else {
                    successfulCount++;
                }
            });
            if (successfulCount > 0) {
                this.log(chalk.green(`\n${successfulCount} device key(s) successfully deleted.`));
            }
            if (failureCount > 0) {
                this.log(chalk.red(`\n${failureCount} device key(s) failed to be deleted.`));
            }
        });
    }
}

File: src/user/device_delete.rs
use std::{convert::TryFrom, path::PathBuf};

use clap::Parser;
use ironoxide::prelude::{BlockingIronOxide, DeviceId};
use yansi::Paint;

use crate::util;

pub const EXAMPLE: &str = "EXAMPLE

    $ ironhide user device-delete 293 5559 312

";

#[derive(Parser)]
#[clap(after_help = EXAMPLE)]
/// Deauthorize a device, so it is no longer able to decrypt your data. Use this to deauthorize devices other than the one you are currently using. To deauthorize your current device, use 'ironhide logout'.
pub struct DeviceDelete {
    /// One or more IDs of the device keys to revoke.
    #[clap(num_args = 1.., required = true)]
    device_ids: Vec<u64>,
    /// Path to location of file which contains keys to use for this operation. Overrides using default key file from
    /// '~/.iron' directory.
    #[clap(value_parser = clap::value_parser!(PathBuf), short, long)]
    keyfile: Option<PathBuf>,
}

impl util::GetKeyfile for DeviceDelete {
    fn get_keyfile(&self) -> Option<&PathBuf> {
        self.keyfile.as_ref()
    }
}

pub fn delete_devices(
    sdk: &BlockingIronOxide,
    DeviceDelete { device_ids, .. }: DeviceDelete,
) -> Result<(), String> {
    for device_id in device_ids.iter() {
        match sdk.user_delete_device(Option::Some(&DeviceId::try_from(*device_id)?)) {
            Ok(dev_id_confirm) => util::println_paint(Paint::green(format!(
                "Successfully deleted device with ID {}",
                dev_id_confirm.id()
            ))),
            Err(e) => util::println_paint(Paint::red(format!(
                "Unable to delete device with ID {} - {}",
                device_id, e
            ))),
        }
    }
    Ok(())
}
